File: src/client/java/com/somefrills/chestui/ColorSelectionMenu.java
package com.somefrills.chestui;

import com.somefrills.misc.MyMapColor;
import com.somefrills.misc.RenderColor;
import com.somefrills.misc.Utils;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.text.Style;
import net.minecraft.util.DyeColor;

public class ColorSelectionMenu extends ChestUI{
    private final RenderColor color;

    public ColorSelectionMenu(ChestUI previousScreen, RenderColor initialColor) {
        super("Color", previousScreen);
        this.color = initialColor;
        rebuild();
    }

    @Override
    protected void build() {
        for (MyMapColor color : MyMapColor.values()) {
            ItemStack stack = new ItemStack(color.getItem());
            int colorHex = color.getColor();
            String name = Utils.capitalizeType(color.name().toLowerCase());
            Utils.setCustomName(stack, Style.EMPTY.withColor(colorHex), name);
            var nbt = new NbtCompound();
            nbt.putInt("color", colorHex);
            Utils.setCustomData(stack, nbt);
            addItem(stack);
        }
    }

    @Override
    protected void onItemClick(ItemStack stack, int button) {
        if (stack.isEmpty()) return;
        int colorHex = Utils.getCustomData(stack).getInt("color").orElse(0xFFFFFF);
        color.set(RenderColor.fromHex(colorHex));
        close();
    }
}

File: src/main/kotlin/com/somefrills/features/misc/glowmob/chestui/ColorSelectionMenu.kt
package com.somefrills.features.misc.glowmob.chestui

import com.somefrills.misc.MyMapColor
import com.somefrills.misc.RenderColor
import com.somefrills.misc.RenderColor.Companion.fromHex
import com.somefrills.utils.TextUtils
import com.somefrills.utils.setCustomData
import com.somefrills.utils.getCustomData
import com.somefrills.utils.setCustomName
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.chat.Style
import net.minecraft.world.item.ItemStack
import java.util.*

class ColorSelectionMenu(previousScreen: ChestUI?, private val color: RenderColor) : ChestUI("Color", previousScreen) {
    init {
        rebuild()
    }

    override fun build() {
        for (color in MyMapColor.entries) {
            val stack = ItemStack(color.item)
            val colorHex = color.hex
            val name = TextUtils.capitalizeType(color.name.lowercase())
            stack.setCustomName(Style.EMPTY.withColor(colorHex), name)
            val nbt = CompoundTag()
            nbt.putInt("color", colorHex)
            stack.setCustomData(nbt)
            addItem(stack)
        }
    }

    override fun onItemClick(stack: ItemStack?, button: Int) {
        if (stack?.isEmpty ?: return) return
        val colorHex = stack.getCustomData()?.getInt("color")?.orElse(0xFFFFFF) ?: 0xFFFFFF
        color.set(fromHex(colorHex))
        onClose()
    }
}
